pysteps/verification/salscores.py

```diff
@@ -0,0 +1,386 @@
+    """Calculate the second parameter of location component for SAL based on :cite:`WPHF2008`.
+
+    Parameters
+    ----------
+    prediction: array-like
+        Array of shape (m,n) with prediction data. NaNs are ignored.
+    observation: array-like
+        Array of shape (m,n)  with observation data. NaNs are ignored.
+    thr_factor: float
+        Factor used to compute the detection threshold as in eq. 1 of :cite:`WHZ2009`.
+        If specified, this is used to identify coherent objects enclosed by the threshold
+        contour thr_factor * quantile95(precip).
+    tstorm_kwargs: dict
+        Optional dictionary containing keyword arguments for the tstorm feature
+        detection algorithm. If None, default values are used.
+        See the documentation of :py:func:`pysteps.feature.tstorm.detection`.
+
+    Returns
+    -------
+    location_2: float
+        The secibd parameter of location component with value between 0 to 1.
+    """
+    maximum_distance = sqrt(
+        ((observation.shape[0]) ** 2) + ((observation.shape[1]) ** 2)
+    )
+    obs_r = (_sal_weighted_distance(observation, thr_factor, tstorm_kwargs)) * (
+        np.nanmean(observation)
+    )
+    forc_r = (_sal_weighted_distance(prediction, thr_factor, tstorm_kwargs)) * (
+        np.nanmean(prediction)
+    )
+    location_2 = 2 * ((abs(obs_r - forc_r)) / maximum_distance)
+    return float(location_2)
+
+
+def _sal_detect_objects(precip, thr_factor, tstorm_kwargs):
+    """This function detects thunderstorms using a multi-threshold approach from :cite:`Feldmann2021`.
+
+    Parameters
+    ----------
+    precip: array-like
+        Array of shape (m,n) containing input data. Nan values are ignored.
+    thr_factor: float
+        Factor used to compute the detection threshold as in eq. 1 of :cite:`WHZ2009`.
+        If specified, this is used to identify coherent objects enclosed by the threshold
+        contour thr_factor * quantile95(precip).
+    tstorm_kwargs: dict
+        Optional dictionary containing keyword arguments for the tstorm feature
+        detection algorithm. If None, default values are used.
+        See the documentation of :py:func:`pysteps.feature.tstorm.detection`.
+
+    Returns
+    -------
+    precip_objects: pd.DataFrame
+        Dataframe containing all detected cells and their respective properties.
+    """
+    if not PANDAS_IMPORTED:
+        raise MissingOptionalDependency(
+            "The pandas package is required for the SAL "
+            "verification method but it is not installed"
+        )
+    if not SKIMAGE_IMPORTED:
+        raise MissingOptionalDependency(
+            "The scikit-image package is required for the SAL "
+            "verification method but it is not installed"
+        )
+    if tstorm_kwargs is None:
+        tstorm_kwargs = dict()
+    if thr_factor is not None:
+        zero_value = np.nanmin(precip)
+        tstorm_kwargs = {
```

Comment:
mindiff is not defined here and thus will use the default "valley depth" of 6 to split contiguous areas - to get results consistent with the threshold approach, this should also be defined as a function of the value distribution (e.g. a quantile)

Reply:
good point! What value could we assign to avoid any splitting of contiguous areas?

Reply:
To subpress all splits, you can put anything bigger than max-min of the entire field. As some people use this with dBZ fields that have negative values, I would compute the valid range explicitly.

Reply:
done in 8789c26 , could you quickly check please?

Reply:
everything looks good to me!

Reply:
Thanks a lot, it looks great. Just a minor suggestion: I suggest we use the 95th quantile instead of maximum for the max_min value as well. 

For plotting: I think using a scatter plot with the L component as the color arg would work. The input can be sal or accum_sal. I'll add it ASAP.

Reply:
To completely disable the splitting, max_min needs to cover the complete range from the minimal to the maximal value, not just the 5th-95th percentile. Perhaps another solution could be to put np.Inf here, similar to the max_ref threshold?

Reply:
Hmm, I think i'm missing something here. The max_min is the same value as mindiff in tstorm (using skim.h_maxima), which is the minimum difference between two maximums in the area to separate it into two objects. 
So if this value is more than maximum (here is max_min +1), no object is detected.

Reply:
Ah, I see, my mistake. Can you try min_diff=np.nanmax(precip)-minref ?

Reply:
Yep we got wrong the mindiff parameter, well spotted!
I ended up with a new set of parameters, see 6b88dcd
The key is to set a very low maxref to practically exclude mindiff and only rely on the minimum threshold.

Example:
![image](https://user-images.githubusercontent.com/11967971/148463986-864b5f57-a3d7-4fb7-8be3-b0cb99731e70.png)

Perhaps not perfect, but close enough to the algorithm in the original papers? what do you think?

Reply:
Thank you, yes. With these thresholds, we are closer to the original papers.
And it is good that users have the option to change the definition of precipitation 'object', so they can change it based on their study areas, precipitation regimes, etc.